akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -91,20 +99,28 @@ private[http2] class Http2ServerDemux(http2Settings: Http2CommonSettings, initia
       override def settings: Http2CommonSettings = http2Settings
       override def isUpgraded: Boolean = upgraded
 
+      def maxConcurrentStreams: Int = http2Settings.maxConcurrentStreams
+
       override protected def logSource: Class[_] = if (isServer) classOf[Http2ServerDemux] else classOf[Http2ClientDemux]
 
       val multiplexer = createMultiplexer(frameOut, StreamPrioritizer.first())
 
+      // Send settings initially based on our configuration. For simplicity, these settings are
+      // enforced immediately even before the acknowledgement is received.
+      // Reminder: the receiver of a SETTINGS frame must process them in the order they are received.
+      private def initialLocalSettings = immutable.Seq(Setting(SettingIdentifier.SETTINGS_MAX_CONCURRENT_STREAMS, maxConcurrentStreams))
+
       override def preStart(): Unit = {
-        if (initialDemuxerSettings.nonEmpty) {
-          debug(s"Applying ${initialDemuxerSettings.length} initial settings!")
-          applySettings(initialDemuxerSettings)
+        if (initialRemoteSettings.nonEmpty) {
+          debug(s"Applying ${initialRemoteSettings.length} initial settings!")
+          applyRemoteSettings(initialRemoteSettings)
         }
 
         pullFrameIn()
         pull(substreamIn)
 
-        multiplexer.pushControlFrame(SettingsFrame(Nil)) // both client and server must send an settings frame as first frame
+        // both client and server must send a settings frame as first frame
```

Comment:
I just realized this comment (and implementation) may be wrong. This is not to be fixed in this PR, but when implementing the client, if we sent the initial settings as the `Upgrade` header then we should not push a `SettingsFrame`.

Reply:
So this is something we'll have to doublecheck when we implement the client negotiating HTTP/2 via the `Upgrade` mechanism :+1: